pensieve/config.py

```diff
@@ -0,0 +1,313 @@
+
+    The `select_expression` of the metric may use Jinja2 template syntax to refer to the
+    aggregation helper functions defined in `mozanalysis.metrics`, like
+        '{{agg_any("payload.processes.scalars.some_boolean_thing")}}'
+    """
+
+    name: str  # implicit in configuration
+    select_expression: str
+    data_source: DataSourceReference
+
+    def resolve(self, spec: "AnalysisSpec") -> mozanalysis.metrics.Metric:
+        select_expression = _metrics_environment.from_string(self.select_expression).render()
+
+        return mozanalysis.metrics.Metric(
+            name=self.name,
+            data_source=self.data_source.resolve(spec),
+            select_expr=select_expression,
+        )
+
+
+@attr.s(auto_attribs=True)
+class MetricsConfiguration:
+    """Contains the metrics configured for each analysis window of an experiment."""
+
+    daily: Set[mozanalysis.metrics.Metric]
+    weekly: Set[mozanalysis.metrics.Metric]
+    overall: Set[mozanalysis.metrics.Metric]
+
+
+@attr.s(auto_attribs=True)
+class MetricsSpec:
+    """Describes the interface for the metrics section in configuration."""
+
+    daily: List[MetricReference] = attr.Factory(list)
+    weekly: List[MetricReference] = attr.Factory(list)
+    overall: List[MetricReference] = attr.Factory(list)
+
+    definitions: Dict[str, MetricDefinition] = attr.Factory(dict)
+
+    @classmethod
+    def from_dict(cls, d: dict) -> "MetricsSpec":
+        params: Dict[str, Any] = {}
+        known_keys = ("daily", "weekly", "overall")
+        for k in known_keys:
+            v = d.get(k, [])
+            if not isinstance(v, list):
+                raise ValueError(f"metrics.{k} should be a list of metrics")
+            params[k] = [MetricReference(m) for m in v]
+        params["definitions"] = {
+            k: _converter.structure({"name": k, **v}, MetricDefinition)
+            for k, v in d.items()
+            if k not in known_keys
+        }
+        return cls(**params)
+
+    @staticmethod
+    def _merge_metrics(
+        user: Iterable[mozanalysis.metrics.Metric], default: Iterable[mozanalysis.metrics.Metric]
+    ) -> Set[mozanalysis.metrics.Metric]:
+        result = set(user)
+        user_names = {m.name for m in user}
+        for m in default:
+            if m.name not in user_names:
+                result.add(m)
+        return result
+
+    def resolve(self, spec: "AnalysisSpec") -> MetricsConfiguration:
+        daily = self._merge_metrics(
+            {ref.resolve(spec) for ref in self.daily}, DEFAULT_METRICS["desktop"]["daily"]
+        )
+        weekly = self._merge_metrics(
+            {ref.resolve(spec) for ref in self.weekly}, DEFAULT_METRICS["desktop"]["weekly"]
+        )
+        overall = self._merge_metrics(
+            {ref.resolve(spec) for ref in self.overall}, DEFAULT_METRICS["desktop"]["overall"]
+        )
+        return MetricsConfiguration(daily, weekly, overall)
```

Comment:
We could optimize this if we want to:

```suggestion
    def resolve(self, spec: "AnalysisSpec") -> MetricsConfiguration:
        def merge(k: str):
            return self._merge_metrics(
                {ref.resolve(spec) for ref in getattr(self, k)}, DEFAULT_METRICS["desktop"][k]
            )

        return MetricsConfiguration(merge("daily"), merge("weekly"), merge("overall"))
```

Reply:
Yeah this is probably worth it; I almost messed this up while copy-pasting.